src/web/services/search.service.ts

```diff
@@ -0,0 +1,277 @@
+import { Injectable } from '@angular/core';
+import { forkJoin, Observable, of } from 'rxjs';
+import { flatMap, map } from 'rxjs/operators';
+import { WebPageEndpoints } from '../types/api-endpoints';
+import {
+  Course,
+  FeedbackSessions,
+  Instructor,
+  InstructorPermissionRole,
+  InstructorPrivilege,
+  Instructors,
+  Student,
+  Students,
+} from '../types/api-output';
+import { Intent } from '../types/api-request';
+import { CourseService } from './course.service';
+import { FeedbackSessionsService } from './feedback-sessions.service';
+import { HttpRequestService } from './http-request.service';
+import { InstructorService } from './instructor.service';
+import { LinkService } from './link.service';
+
+/**
+ * Handles the logic for search.
+ */
+@Injectable({
+  providedIn: 'root',
+})
+export class SearchService {
+
+  constructor(
+    private instructorService: InstructorService,
+    private httpRequestService: HttpRequestService,
+    private feedbackSessionService: FeedbackSessionsService,
+    private courseService: CourseService,
+    private linkService: LinkService) {
+  }
+
+  searchAdmin(searchKey: string): Observable<AdminSearchResult> {
+    return forkJoin(
+      this.getStudents(searchKey).pipe(
+        map((students: Students) => students.students),
+        flatMap((studentsArray: Student[]) =>
+                studentsArray.length !== 0
+                  ? forkJoin(studentsArray.map((student: Student) => this.createStudentAccountSearchResult(student)))
```

Comment:
If you have time, consider how many redundant requests you may have send through the request.

Reply:
The two ways I can think of: 
1. Do some sort of caching for common routes like courses, so we dont make the same call for the course ids
2. Open up new endpoints, such as on privilege, so we can pass it an array of ids instead of calling it one by one.

Reply:
Fixed the redundant calls issues by aggregating all the course ids in a set, to generate a map of course ids to the various api responses